### PR TITLE
Some suggesting changed to add_noise

### PR DESCRIPTION
**Pull request by ycasg**
I think that there was an error in the power/bin calculation. See
suggested change. I also simplified the code using different Pulse
helper functions.

### CODE REVIEW

**Comment by DanHickstein** [comment]
I like the way that you have phrased the calculation of the number of photons per bin (now photons/bin/sec). Does it actually change the number of photons added? It looks like maybe it changes by 2*pi...

**Comment by ycasg** [comment]
There was one big change, which is the power/bin. I think that a division should be a multiplication:
```python
energy_per_bin = np.abs(self.AW)**2/size_of_bins * 1e-12
``` 
to
```python
power_per_bin = np.abs(self.AW)**2 * size_of_bins  # [W/Hz]  * [Hz]
```
This should be verified (or maybe you already did, and the division is correct?) If done correctly, the integrated result should be the average (CW) power for a CW input pulse.

**Comment by DanHickstein** [comment]
Oh, this is a bit confusing. Do we want to end up with photons/sec or the actual number of photons in that frequency bin for the pulse?

**Comment by ycasg** [comment]
In the time domain, the number of photons would be correct. The field AT is in rtWatts, and if you multiply by the time-domain width dT you get energy per time bin. This could be converted to photons, and shot noise could be added. 

In the frequency domain, AW is in something like rtWatts/rtHz. What it is exactly I not 100 % sure -- the nonlinearity is calculated in the time domain, and the algorithm is therefore set up to be in physical units there, but not necessarily correct in the frequency domain. Specifically, the FFT performs a normalization by 1/N in the calculation of the FFT (or IFFT -- I'm not totally sure which.) The end result is that you should check the scaling factors using a CW pulse to make sure that things are what you expect. 

Specifically, if you have a 1 watt average power CW "pulse", then there should be 1 W in a single bin in AW. I haven't checked that this is the case.

**Comment by ycasg** [comment]
The more I think about this, the more I feel like shot noise should be added in the time domain, not the frequency domain. At a simplified level, shot noise is the statistical uncertainty in the number of photons detected in an amount of time. When I think about the CW case, I expect to see a little fuzz on $P(t)$ proportional to $P(t) \Delta t $ . Taking the Fourier transform, I think that this will map to a little bit of noise power at all optical frequencies.

Because it is essentially counting noise, I don't know how to calculate shot noise directly from the frequency domain, or if this is possible / meaningful. The number "photons counted in some amount of something" doesn't exist in the frequency domain, only "photons per unit time per optical bandwidth." 

Two possibilities are (1) think really hard about is and (2) try a time domain version of the noise calculation. It's probably good to do (1), but (2) is easy. It also would most likely increase the amount of broadband noise, which you found to have a larger impact on coherence.

**Comment by DanHickstein** [comment]
Yes, the more that I think about it, including noise as sqrt(photons) per bin in the frequency domain seems to miss the point: it only includes noise at frequencies already present in the pulse, so I would not expect the output temporal waveform to look all that different.

I think that it might be pretty straightforward to include the noise in the time domain. We can just calculate the energy in each bin of `AT**2 * bin_size`, since we would have [W] * [s] = J. 

The weird part is that, to get to number of photons, then we need to make some assumption about the photon energy in that time bin, and we don't really have any way of knowing this, since that time bin just has one value of the E-field. However, in the case of a transform limited Gaussian pulse, then all frequencies are equally present at every point in time (right?), so can we simply say that the photon energy is the average photon energy of the pulse. Then, we could include sqrt(N) photons of noise in each energy bin. I wonder if this will magically become identical to including 1 photon per bin in the frequency domain, as several of the papers indicated.

If the pulse is chirped, then it does not seem like a good approximation to tread the photon energy as constant for each time bin. For now, perhaps we could just include in the docstring that the method is only strictly correct for transform-limited pulses. Ultimately, it might be possible to generate something like a FROG spectrogram, that allows us to calculate an average photon energy as a function of time across the pulse. Then, we could more accurately assign the photon energy for each temporal bin in the case of chirped or otherwise complicated pulses.

File: src/pynlo/light/PulseBase.py [hunk 1]
```diff
@@ -636,13 +636,15 @@ def add_noise(self, noise_type='sqrt_N_freq'):
...
         
-        # This is all to get the number of photons in each frequency bin:
-        size_of_bins = np.gradient(self.W_THz)
-        energy_per_bin = np.abs(self.AW)**2/size_of_bins * 1e-12
-    
-        h = 6.62607004e-34
-        photon_energy = h * self.W_THz/(2*np.pi) * 1e12
-        photons_per_bin = energy_per_bin/photon_energy
+        # This is all to get the number of photons/second in each frequency bin:
+        size_of_bins = self.dF_mks                          # Bin width in [Hz]
```
Comment (DanHickstein) [inline src/pynlo/light/PulseBase.py]: since the bin size has been changed to Hz, rather than THz, the factor of 1e12 should be removed from line 661

**Comment by DanHickstein** [comment]
I tried the examples/SSFM_coherence.py with both `noise_type == 'sqrt_N_freq'` and `noise_type == 'one_photon_freq'`. Both seem to work fine, as long as the factor of 1e12 is removed from line 661 as mentioned above.

Once that change is made, I think that this can be merged. Of course, we should really implement the noise in the time domain...

**Pull request merged**